lib/Echidna/Processor.hs

```diff
@@ -0,0 +1,174 @@
+{-# LANGUAGE LambdaCase #-}
+
+module Echidna.Processor where
+
+import Control.Arrow              (second)
+import Control.Lens
+import Control.Monad.IO.Class     (MonadIO(..))
+import Control.Exception          (Exception)
+import Control.Monad.Catch        (MonadThrow(..))
+import Data.Aeson                 (decode, Value(..))
+import Data.Text                  (Text, pack, unpack)
+import System.Directory           (findExecutable)
+import System.Process             (StdStream(..), readCreateProcessWithExitCode, proc, std_err)
+import System.Exit                (ExitCode(..))
+import GHC.Word                   (Word32)
+
+import qualified Data.ByteString.Lazy.Char8 as BSL
+import qualified Data.HashMap.Strict as M
+
+import Echidna.ABI (hashSig)
+
+-- | Things that can go wrong trying to run a processor. Read the 'Show'
+-- instance for more detailed explanations.
+data ProcException = ProcessorFailure String String
+                   | ProcessorNotFound String
+
+instance Show ProcException where
+  show = \case
+    ProcessorFailure p e     -> "Error running " ++ p ++ ": " ++ e
+    ProcessorNotFound p      -> "Cannot find processor " ++ p
+
+instance Exception ProcException
+
+filterResults :: Maybe String -> [(Text, [Text])] -> [Word32] 
+filterResults (Just c) rs = 
+  case lookup (pack c) rs of
+    Nothing -> filterResults Nothing rs
+    Just s -> map hashSig s 
+
+filterResults Nothing rs = concatMap (map hashSig . snd) rs
+
+data SlitherInfo = PayableInfo (Text, [Text]) 
+                 | ConstantFunctionInfo (Text, [Text])
+                 | AssertFunction (Text, [Text])
+                 | ConstantValue (Text, Text)
+                 | GenerationGraph (Text, Text, [Text]) deriving (Show)
+
+filterPayable :: [SlitherInfo] -> [(Text, [Text])]
+filterPayable = map g . filter f
+  where f (PayableInfo _) = True
+        f _               = False
+        g (PayableInfo i) = i
+        g _               = error "fail in filterPayable"
+
+filterAssert :: [SlitherInfo] -> [(Text, [Text])]
+filterAssert = map g . filter f
+  where f (AssertFunction _) = True
+        f _               = False
+        g (AssertFunction i) = i
+        g _               = error "fail in filterAssert"
+
+filterConstantFunction :: [SlitherInfo] -> [(Text, [Text])]
+filterConstantFunction = map g . filter f 
+  where f (ConstantFunctionInfo _) = True
+        f _               = False
+        g (ConstantFunctionInfo i) = i
+        g _               = error "fail in filterConstantFunction"
+
+filterGenerationGraph :: [SlitherInfo] -> [(Text, Text, [Text])]
+filterGenerationGraph = map g . filter f
+  where f (GenerationGraph _) = True
+        f _                   = False
+        g (GenerationGraph i) = i
+        g _                   = error "fail in filterGenerationGraph"
+
+-- Slither processing
+runSlither :: (MonadIO m, MonadThrow m) => FilePath -> [String] -> m [SlitherInfo]
```

Comment:
I'm not sure about the return type here. It looks like it provides a little value in this form and has to be further processed in `prepareForTest`. What about returning a record similar to:
```haskell
data SlitherInfo =
  { payableInfo :: [(Text, [Text])]
  , constantFunctionInfo :: [(Text, [Text])]
  , assertFunction :: [(Text, [Text])]
  , constantValue :: [(Text, Text)]
  , generationGraph :: [(Text, Text, [Text])]
  }
```
or even organize it by function if possible

Reply:
It is easier to parse each constructor and the filter them. It is a pain to process JSON in Haskell...